components/esp_modem/examples/modem_tcp_client/components/extra_tcp_transports/include/mbedtls_wrap.hpp

```diff
@@ -41,12 +43,33 @@ class Tls {
     mbedtls_entropy_context entropy_{};
     virtual void delay() {}
 
+    bool set_session();
+    bool get_session();
+    void reset_session();
+    bool is_session_loaded();
+
 private:
     static void print_error(const char *function, int error_code);
     static int bio_write(void *ctx, const unsigned char *buf, size_t len);
     static int bio_read(void *ctx, unsigned char *buf, size_t len);
     int mbedtls_pk_parse_key( mbedtls_pk_context *ctx,
                               const unsigned char *key, size_t keylen,
                               const unsigned char *pwd, size_t pwdlen);
+    struct unique_session {
+        unique_session()
+        {
+            ::mbedtls_ssl_session_init(&s);
+        }
+        ~unique_session()
+        {
+            ::mbedtls_ssl_session_free(&s);
+        }
+        mbedtls_ssl_session *ptr()
+        {
+            return &s;
+        }
+        mbedtls_ssl_session s;
+    };
+    std::unique_ptr<unique_session> session_;
```

Comment:
Reading the code in the current state it seems that we only have the `std::unique_ptr` because the session isn't necessarily loaded, it seems that we could use a `std::optional` instead.

Reply:
Yes, the `session_` could be used but doesn't necessarily have to be used, so for the entire lifetime could remain `nullptr`.
And yes, I use `unique_ptr` only destroy the potentially allocated object, if it's used.

I think we could also use `std::optional`, but it seems to me that it would be easier to use a raw pointer and add a `delete session_` to the Tls destructor?

Reply:
With your current design, with the `unique_session` type, the std::optional adds the same effect since the destructor would be called on TLS destructor. The raw pointer cleanup would work also, but I prefer to keep a type dedicated to lifetime control in such cases, just a matter of taste though. The solution with `unique_ptr` is ok, just that optional reflects better the meaning.

Reply:
Have reworked to use `std::optional`, could you please take a quick look?

Reply:
Reverted to the `unique_session` per discussion with @euripedesrocha since this is being added to the base class and using session tickets doesn't seem to be a very common usecase for everyone.